design/design-doc-rbac-manager.md

```diff
@@ -0,0 +1,706 @@
+
+* A provider's service account the access it needs to reconcile the managed
+  resources it defines.
+* Crossplane's service account access to compose the managed resources a
+  provider defines.
+* Crossplane's service account access to reconcile any defined composite
+  resources and composite resource claims.
+
+Crossplane should also create (but not bind) RBAC roles that grant:
+
+* Platform operators access to interact with the managed resources a provider
+  defines.
+* Platform operators access to interact with the composite resources and
+  composite resource claims they define.
+* Platform consumers access to interact with the set of composite resource
+  claims their platform operator(s) deem appropriate.
+
+Finally, it should be possible to opt-out of RBAC role management. This allows
+Crossplane to be run without any special privileges (i.e. without the ability to
+escalate its own privileges) at the expense of deferring RBAC role management to
+platform operators.
+
+## Proposal
+
+This document proposes that Crossplane introduce an 'RBAC manager'. The RBAC
+manager would run as a distinct pod, and would be responsible for creating and
+binding the RBAC roles Crossplane and its users need. This allows Crossplane and
+its providers run as service accounts that are limited to a specific set of
+resources and verbs (e.g. create, watch, etc). The RBAC manager would also run
+as a service account limited to a specific set of resources and verbs, but would
+be granted the special `escalate` and `bind` on the `Role` and `ClusterRole`
+resources, thus allowing it to grant access that it does not have.
+
+Privilege escalation is an unfortunate by-product of a controller manager that
+may dynamically define new custom resources. Centralising the ability to
+escalate privileges into the RBAC manager has two nice properties:
+
+* The platform operator can choose not to deploy the RBAC manager, and can
+  instead manually curate RBAC roles - e.g. to manually create and bind RBAC
+  roles for each provider and composite resource.
+* The attack surface required to exploit Crossplane is reduced. Fewer Crossplane
+  controllers run with the ability to escalate privileges.
+
+### Managed RBAC ClusterRoles
+
+The RBAC manager will be responsible for the management of several cluster
+roles, using [`ClusterRole` aggregation][Aggregated ClusterRoles] to form a set
+of cluster roles that may change over time as providers and composite resources
+are added to and removed from the API server. The RBAC manager would manage the
+following cluster roles. Note that these roles are aggregated - they consist of
+several more granular roles that are not covered here. **The following examples
+are 'flattened' aggregations represented as a single fixed `ClusterRole` to
+demonstrate specifically what permissions each role will have.**
+
+The Crossplane `ClusterRole` is bound to the service account the Crossplane pod
+runs as. It grants the following rules:
```

Comment:
We will also need the following permissions for the existing package manager, as well as the new package manager (which we may be able to be paired down after removal of the existing):
```yaml
rules:
  - apiGroups: ["extensions", "apps"]
    resources: ["deployments"]
    verbs: ["get", "list", "create", "update", "patch", "delete", "deletecollection"]
  - apiGroups: ["batch", "extensions"]
    resources: ["jobs"]
    verbs: ["get", "list", "create", "update", "patch", "delete", "deletecollection"]
  - apiGroups: [""]
    resources: ["secrets"]
    verbs: ["get", "list", "create", "update", "patch", "delete"]
  - apiGroups: [""]
    resources: ["pods"]
    verbs: ["get", "list"]
  - apiGroups: [""]
    resources: ["pods/log"]
    verbs: ["get"]
```
However, these are only needed for a `Role` and can be bound with a `RoleBinding` as the package manager only operates in the namespace Crossplane is running in. Note that these are the [same privileges the package manager has in hosted mode](https://github.com/crossplane/crossplane/blob/master/cluster/charts/crossplane-controllers/templates/package-manager-host-role.yaml). The package manager also will need cluster-scoped permissions, but that is included in the `ClusterRole` you defined below :+1:

Reply:
Yeah, this occurred to me this morning as well. I like your idea of keeping the existing package manager as a distinct deployment, rolling the new package manager into the core deployment, and granting the core deployment the access it needs (jobs, deployments, pods, etc).

Reply:
I've updated this role; let me know how it looks. I removed a few things like `deletecollection` and `list` that I suspect aren't really used, but I could be wrong.